File: Warming_Beam.py
import numpy as np
import os
import matplotlib.pyplot as plt

class WarmingBeam:
    def __init__(self, mx, CFL, a, folder):
        self.mx = mx
        self.dx = 1.0 / mx
        self.CFL = CFL
        self.a = a
        self.dt = CFL * self.dx / a
        self.folder = folder

        self.u_prev = np.zeros(mx + 1)
        self.u_new = np.zeros(mx + 1)

        for i in range(mx + 1):
            x = i * self.dx
            self.u_prev[i] = self.initial_condition(x)
            self.u_new[i] = self.u_prev[i]

        if not os.path.exists(self.folder):
            os.makedirs(self.folder)
        sim.save_data(0)

    def initial_condition(self, x):
        x -= 0.5  # Center the domain
        return 1.0 if -0.25 <= x <= 0.25 else 0.0

    def forward(self):
        u = self.u_prev.copy()
        c = self.CFL

        u_m1 = np.roll(u, 1)   # circshift(u, 1)
        u_m2 = np.roll(u, 2)   # circshift(u, 2)

        self.u_new = u - 0.5 * c * (3*u - 4*u_m1 + u_m2) \
                    + 0.5 * c**2 * (u - 2*u_m1 + u_m2)

        self.u_prev = self.u_new.copy()


    def save_data(self, t):
        filename = os.path.join(self.folder, f"t_{t:.1f}.txt")
        with open(filename, 'w') as f:
            for i in range(self.mx + 1):
                f.write(f"{i * self.dx:.10e} {self.u_new[i]:.10e}\n")

    def visualize(self):
        plt.figure(figsize=(10, 6))
        colors = ['r', 'g', 'b']
        save_times = [0.1, 1.0, 10.0]

        for i, t in enumerate(save_times):
            filename = os.path.join(self.folder, f"t_{t:.1f}.txt")
            if not os.path.exists(filename):
                print(f"File not found: {filename}")
                continue

            data = np.loadtxt(filename)
            x, u = data[:, 0], data[:, 1]
            plt.plot(x, u, label=f't = {t}', color=colors[i % len(colors)])

        plt.xlabel('x')
        plt.ylabel('u')
        plt.title('Warming-Beam Scheme Results')
        plt.legend()
        plt.grid(True)
        plt.tight_layout()
        plt.show()


if __name__ == "__main__":
    sim = WarmingBeam(mx=100, CFL=0.5, a=1.0, folder="output")

    t = 0.0
    save_times = [0.1, 1.0, 10.0]
    save_index = 0
    step = 0
    while t <= save_times[-1] + 1e-6:
        if save_index < len(save_times) and abs(t - save_times[save_index]) < 1e-8:
            print(f"Saving data at t = {t:.1f}")
            sim.save_data(t)
            save_index += 1

        sim.forward()
        t += sim.dt
        step += 1

    sim.visualize()

File: src/Warming_Beam.cpp
#include "Warming_Beam.h"
#include <fstream>
#include <sstream>
#include <iostream>
#include <vector>
#include <string>
#include <cmath>

void WarmingBeam::Initialize(int mx_, long double CFL_, long double a_, std::string folder_) {
    mx = mx_;
    dx = 1.0L / mx;
    CFL = CFL_;
    a = a_;
    dt = CFL * dx / a;

    u_prev.resize(mx + 1);
    u_new.resize(mx + 1);
    
    for (int i = 0; i <= mx; i++) {
        long double x = i * dx;
        u_prev[i] = Initial_condition(x);
        u_new[i] = u_prev[i];
    }

    folder = folder_;
}

long double WarmingBeam::Initial_condition(long double x) {
    x -= 0.5L;  // Center the domain at 0

    if (x >= -0.25L && x <= 0.25L) return 1.0L;
    return 0.0L;
}

void WarmingBeam::Forward() {
    std::vector<long double> u_temp(mx + 1);

    for (int i = 1; i <= mx; i++) {
        // Circular shift implementation
        int i_minus_1 = (i - 1 + mx + 1) % (mx + 1);
        int i_minus_2 = (i - 2 + mx + 1) % (mx + 1);

        u_temp[i] = u_prev[i]
            - 0.5L * CFL * (3.0L * u_prev[i] - 4.0L * u_prev[i_minus_1] + u_prev[i_minus_2])
            + 0.5L * CFL * CFL * (u_prev[i] - 2.0L * u_prev[i_minus_1] + u_prev[i_minus_2]);
    }
    u_temp[0] = u_new[mx];
    u_prev = u_temp;
    u_new = u_temp;
}

void WarmingBeam::Data(int step) {
    std::ostringstream filename;
    filename << folder << "/output_step_" << step << ".txt";

    std::ofstream file(filename.str());
    // Use std::scientific to preserve precision
    file << std::scientific;
    for (int i = 0; i <= mx; i++) {
        file << i * dx << " " << u_new[i] << "\n";
    }
    file.close();
}